rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+
+Alternative Strategy for Enforcing that Type Versions Match
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Sometimes the recommended strategy would interfere with features of the middleware that allow for more complex type compatibility rules, or otherwise interferes with the function of the underlying middleware.
+In these cases, it is appropriate to not take the recommended strategy and delegate the matching process and notification entirely to the middleware.
+
+However, in order to increase compatibility between rmw implementations, it is recommended to fulfill the recommended approach whenever possible, even if the type name is not used in determining if two endpoints will match, i.e. in the case that the underlying middleware does something more sophisticated to determine type compatibility but ignores the type name itself.
+
+This recommendation is particularly useful in the case where a DDS based ROS 2 endpoint is talking to another DDS-XTypes based ROS 2 endpoint.
+The DDS-XTypes based endpoint then has a chance to "gracefully degrade" to interoperate with the basic DDS based ROS 2 endpoint.
+This would not be the case if the DDS-XTypes based ROS 2 endpoint did not include the type version hash in the type name, as is suggested with the recommended strategy.
+
+.. TODO::
+
+    We need to confirm whether or not having a different type name prevents DDS-XTypes from working properly.
+    We've done some experiments, but we need to summarize the results and confirm the recommendation in the REP specifically geared towards DDS.
+
+Notifying the ROS Client Library
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+One potential downside to delegating type matching to the rmw implementation is that detecting the mismatch is more complicated.
+If ROS 2 is to provide users a warning that two endpoints will not communicate due to their types not matching, it requires there to be a way for the middleware to notify the ROS layer when a topic is not matched due to the type incompatibility.
+As some of the following sections describe, it might be that the rules by which the middleware decides on type compatibility are unknown to ROS 2, and so the middleware has to indicate when matches are and are not made.
+
+If the middleware just uses the type name to determine compatibility, then the rmw implementation can just check the type version hash, and if they do not match between endpoints then the rmw implementation can notify ROS 2, and a warning can be produced.
+
+Either way, to facilitate this notice, the ``rmw_event_type_t`` shall be extended to include a new event called ``RMW_EVENT_OFFERED_TYPE_INCOMPATIBLE``.
+Related functions and structures will also be updated so that the event can be associated with specific endpoints.
+
+.. TODO::
+
+    It's not clear how we will do this just now, since existing "QoS events" lack a way to communicate this information, I (wjwwood) think.
```

Comment:
QoS events lack a way to communicate which information? The specific endpoint? The current events, provided in callbacks, do carry event-specific struct payloads that are filled out by the RMW implementation. The current ones are all simple number types at the moment, but that's not a hard requirement. I don't see a reason why `rmw_topic_endpoint_info_t` couldn't be included in the callback payload even, or just the GID or something from it.